Product/Production/Services/PatientDiscoveryCore/src/main/java/gov/hhs/fha/nhinc/patientdiscovery/parser/PRPAIN201305UV02Parser.java

```diff
@@ -0,0 +1,134 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package gov.hhs.fha.nhinc.patientdiscovery.parser;
+
+import java.util.List;
+import org.apache.log4j.Logger;
+import org.hl7.v3.II;
+import org.hl7.v3.PRPAIN201305UV02;
+import org.hl7.v3.PRPAMT201306UV02LivingSubjectId;
+
+/**
+ * This is a utility class that parses the PRPAIN201305UV02 message and extract patientId and participantObjectId from
+ * it.
+ *
+ * @author tjafri
+ */
+public class PRPAIN201305UV02Parser {
+
+    private static final Logger LOG = Logger.getLogger(PRPAIN201305UV02Parser.class);
+
+    public static II getPatientId(PRPAIN201305UV02 request) {
+        LOG.info("getPatientId()");
+        II livingSubjectId = null;
+
+        if (request != null && request.getControlActProcess() != null
+            && request.getControlActProcess().getQueryByParameter() != null
+            && request.getControlActProcess().getQueryByParameter().getValue() != null
+            && request.getControlActProcess().getQueryByParameter().getValue().getParameterList() != null
+            && request.getControlActProcess().getQueryByParameter().getValue().getParameterList()
+            .getLivingSubjectId() != null) {
+
+            List<PRPAMT201306UV02LivingSubjectId> ids = request.getControlActProcess().getQueryByParameter().getValue()
+                .getParameterList().getLivingSubjectId();
+
+            if (ids.size() == 1) {
+                livingSubjectId = getLivingSubjectId(ids.get(0));
+            } else if (ids.size() > 1) {
+                livingSubjectId = getLivingSubjectIdFromAuthorOrPerformerValue(request, ids);
+            }
+        } else {
+            LOG.error("PatientId doesn't exist in the received PRPAIN201305UV02 message");
+        }
+
+        return livingSubjectId;
+    }
+
+    public static String getQueryId(PRPAIN201305UV02 request) {
+        LOG.info("getQueryId()");
+        String oid = null;
+
+        if (request != null
+            && request.getControlActProcess() != null
+            && request.getControlActProcess().getQueryByParameter() != null
+            && request.getControlActProcess().getQueryByParameter().getValue() != null
+            && request.getControlActProcess().getQueryByParameter().getValue().getQueryId() != null) {
+
+            oid = request.getControlActProcess().getQueryByParameter().getValue().getQueryId().getExtension();
+        }
+        return oid;
+    }
+
+    private static II getLivingSubjectIdFromAuthorOrPerformerValue(PRPAIN201305UV02 request,
+        List<PRPAMT201306UV02LivingSubjectId> ids) {
```

Comment:
"ids" is dereferenced in the if, and the else, without a null check.  Is "ids" guaranteed to be non-null (also, non-empty, since in the "else" we call .get(0))?